keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,815 @@
+
+- if all app containers are failed and the restart policy is "never" then terminate the pod. 
+If failure logs are concerned in this case (considering the logging container is sidecar) then the sidecar container will only be terminated when all app containers are in the failed state, so it must have captured the failure logs.
+
+The following are existing behaviors just mentioned here for clarity:
+
+- if all application containers exited successfully and the restart policy of the pod is "always" then restart the containers.
+
+- if all app containers are failed and the restart policy is "onfailure", then kubelet will keep trying to restart containers, and no point in terminating the pod
+
+Since container annotations are not supported to be added by an end user, We need to use pod annotation to specify application containers.
+
+There are few options for annotations:
+
+1. Specifying a list of keystone or sidecar containers in one annotation
+   e.g. `alpha.node.kubernetes.io/sidecars:{list of sidecar containers}` or
+        `alpha.node.kubernetes.io/keystone:{list of keystone containers}`
+2. One annotation per container
+   e.g. `alpha.sidecars.kubernetes.io/<container name>:string`
+
+The motivation for one annotation per container is easy translation to api fields later on and it can be used for the sidecar ordering proposals.
+
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If we choose to specify sidecars container in pod annotation then the user can add all containers in a pod as sidecars and kubelet will terminate the pod. This can be thought of as intended behavior and misuse of annotation than risk. In future, a validation can be added for it.
+
+If we choose to specify keystone containers in pod annotation then even if a user specifies all containers as keystone, the pod will be keep running. And if the user has not specified any annotation kubelet will treat all containers as same as before.
```

Comment:
All containers = keystone means no behaviour difference from the current kubelet lifecycle, right?

Reply:
Hmm, no imho this is unspecified behavior because killing all keystones because one has failed can be unexpected. See next comment (limit to 1 keystone).
However all containers as sidecars means no keystone, and therefore current kubelet lifecycle.